fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/mixin/event/interaction/MixinServerPlayerInteractionManager.java

```diff
@@ -81,4 +86,28 @@ public void interactItem(ServerPlayerEntity player, World world, ItemStack stack
 			return;
 		}
 	}
+
+	@Inject(at = @At(value = "INVOKE", target = "Lnet/minecraft/block/Block;onBreak(Lnet/minecraft/world/World;Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/block/BlockState;Lnet/minecraft/entity/player/PlayerEntity;)V"), method = "tryBreakBlock", locals = LocalCapture.CAPTURE_FAILHARD, cancellable = true)
+	private void breakBlock(BlockPos pos, CallbackInfoReturnable<Boolean> cir, BlockState state, BlockEntity entity, Block block) {
+		ActionResult result = BlockBreakEvents.BEFORE.invoker().beforeBlockBreak(this.world, this.player, pos, state, entity, block);
+
+		if (result == ActionResult.FAIL) {
+			BlockPos cornerPos = pos.add(-1, -1, -1);
+
+			for (int x = 0; x < 3; x++) {
+				for (int y = 0; y < 3; y++) {
+					for (int z = 0; z < 3; z++) {
+						this.player.networkHandler.sendPacket(new BlockUpdateS2CPacket(world, cornerPos.add(x, y, z)));
```

Comment:
Instead of doing this loop, add a `Cancellation` sub functioal interface that allow mods to handle custom server-side block break cancellation logic. In this setup, you only consider vanilla redstone but disregard other mods that can have a bigger volume of influence.

https://github.com/FabricMC/fabric/pull/980/files#r463989517

Preferably, this loop should be moved to an event handler listening to the cancellation event than left in this mixin body.

Reply:
Mods could just send the packet themselves by listening to the event